File: script/make/protocol/lua/RankWorldProtocol.lua
RankWorldProtocol = {}

function RankWorldProtocol.encode(offset, protocol, data)
    if protocol == 19201 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19202 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19203 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19204 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 19205 then
        local offset = offset
        local table = {}
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function RankWorldProtocol.decode(offset, protocol, data)
    if protocol == 19201 then
        local offset = offset
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId}
            data[dataIndex] = rank
        end
        return data
    elseif protocol == 19202 then
        local offset = offset
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local otherLevel = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local otherClasses = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = otherLevel, classes = otherClasses}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            data[dataIndex] = rank
        end
        return data
    elseif protocol == 19203 then
        local offset = offset
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local otherLevel = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local otherClasses = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local otherSex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = otherLevel, classes = otherClasses, sex = otherSex}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            data[dataIndex] = rank
        end
        return data
    elseif protocol == 19204 then
        local offset = offset
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local otherLevel = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local otherClasses = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local otherSex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- VIP等级
            local otherVipLevel = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = otherLevel, classes = otherClasses, sex = otherSex, vipLevel = otherVipLevel}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            data[dataIndex] = rank
        end
        return data
    elseif protocol == 19205 then
        local offset = offset
        -- 
        local data = {}
        local dataLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 类型
            local type = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 排名
            local order = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 键
            local key = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 值
            local value = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 时间
            local time = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 名字
            local name = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(name)
            -- 服务器ID
            local serverId = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 
            -- 等级
            local otherLevel = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 职业
            local otherClasses = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 性别
            local otherSex = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- VIP等级
            local otherVipLevel = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- 头像
            local otherAvatar = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local other = {level = otherLevel, classes = otherClasses, sex = otherSex, vipLevel = otherVipLevel, avatar = otherAvatar}
            -- object
            local rank = {type = type, order = order, key = key, value = value, time = time, name = name, serverId = serverId, other = other}
            data[dataIndex] = rank
        end
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: src/module/rank/rank_world_protocol.erl
-module(rank_world_protocol).
-export([read/2, write/2]).
-include("rank.hrl").


-spec read(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
read(19201, <<>>) ->
    {ok, []};

read(19202, <<>>) ->
    {ok, []};

read(19203, <<>>) ->
    {ok, []};

read(19204, <<>>) ->
    {ok, []};

read(19205, <<>>) ->
    {ok, []};

read(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec write(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
write(19201, List) ->
    ListBinary = protocol:write_list(fun(#rank{type = Type, order = Order, key = Key, value = Value, time = Time, name = Name, server_id = ServerId}) -> <<Type:16, Order:64, Key:64, Value:64, Time:32, (byte_size(Name)):16, (Name)/binary, ServerId:16>> end, List),
    {ok, protocol:pack(19201, <<ListBinary/binary>>)};

write(19202, List) ->
    ListBinary = protocol:write_list(fun(#rank{type = Type, order = Order, key = Key, value = Value, time = Time, name = Name, server_id = ServerId, other = {Level, Classes}}) -> <<Type:16, Order:64, Key:64, Value:64, Time:32, (byte_size(Name)):16, (Name)/binary, ServerId:16, Level:16, Classes:8>> end, List),
    {ok, protocol:pack(19202, <<ListBinary/binary>>)};

write(19203, List) ->
    ListBinary = protocol:write_list(fun(#rank{type = Type, order = Order, key = Key, value = Value, time = Time, name = Name, server_id = ServerId, other = {Level, Classes, Sex}}) -> <<Type:16, Order:64, Key:64, Value:64, Time:32, (byte_size(Name)):16, (Name)/binary, ServerId:16, Level:16, Classes:8, Sex:8>> end, List),
    {ok, protocol:pack(19203, <<ListBinary/binary>>)};

write(19204, List) ->
    ListBinary = protocol:write_list(fun(#rank{type = Type, order = Order, key = Key, value = Value, time = Time, name = Name, server_id = ServerId, other = {Level, Classes, Sex, VipLevel}}) -> <<Type:16, Order:64, Key:64, Value:64, Time:32, (byte_size(Name)):16, (Name)/binary, ServerId:16, Level:16, Classes:8, Sex:8, VipLevel:8>> end, List),
    {ok, protocol:pack(19204, <<ListBinary/binary>>)};

write(19205, List) ->
    ListBinary = protocol:write_list(fun(#rank{type = Type, order = Order, key = Key, value = Value, time = Time, name = Name, server_id = ServerId, other = {Level, Classes, Sex, VipLevel, Avatar}}) -> <<Type:16, Order:64, Key:64, Value:64, Time:32, (byte_size(Name)):16, (Name)/binary, ServerId:16, Level:16, Classes:8, Sex:8, VipLevel:8, Avatar:8>> end, List),
    {ok, protocol:pack(19205, <<ListBinary/binary>>)};

write(Protocol, Data) ->
    {error, Protocol, Data}.


